### PR TITLE
[Merged by Bors] - Fix label macro for types with generics

### LINKED ISSUE #1497: Typed SystemLabels regress generic labeling
## Bevy version

bevy\main

## Operating system & version

Win10

## What you did

Updating to main branch with #1473 disallows creating labels that are generic over a type.

## What you expected to happen

When stringly typed, I could have labels that were unique per concrete type over a generic. I haven't found a workaround to do the same with typed labels in their current form. I'd like to have generic labels so that users of my plugin can have multiple concrete types of my generic system active at the same time.

## What actually happened

Attempting to derive SystemLabel for an enum with generic type annotations results in an error, probably because the derive macro doesn't support generics in the signature. Using an enum without generics results in a runtime error due to a non-unique SystemLabel.

### PR DESCRIPTION
**Pull request by TheRawMeatball**
Fixes #1497

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

**Comment by aevyrie** [comment]
I switched to this PR branch, but am still seeing a type error with the following code:

```rust
#[derive(Debug, Hash, PartialEq, Eq, Clone, SystemLabel)]
pub enum BoundingSystem<T> {
    UpdateBoundVols(PhantomData<T>),
    UpdateDebugMeshes(PhantomData<T>),
}
```

errors:

```shell
error[E0107]: wrong number of type arguments: expected 1, found 0
  --> src\lib.rs:17:10
   |
17 | pub enum BoundingSystem<T> {
   |          ^^^^^^^^^^^^^^ expected 1 type argument

error[E0107]: wrong number of type arguments: expected 0, found 1
  --> src\lib.rs:17:25
   |
17 | pub enum BoundingSystem<T> {
   |                         ^ unexpected type argument

error[E0277]: the trait bound `BoundingSystem<T>: bevy::prelude::SystemLabel` is not satisfied
  --> src\lib.rs:36:28
   |
36 |                     .label(BoundingSystem::UpdateBoundVols(PhantomData::<T>::default())),
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `bevy::prelude::SystemLabel` is 
not implemented for `BoundingSystem<T>`
```

File: crates/bevy_ecs/macros/src/lib.rs [hunk 1]
```diff
@@ -500,8 +500,10 @@ fn derive_label(input: DeriveInput, label_type: Ident) -> TokenStream2 {
     };
     let crate_path: Path = syn::parse(path_str.parse::<TokenStream>().unwrap()).unwrap();
 
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
     quote! {
-        impl #crate_path::#label_type for #ident {
+        impl #impl_generics #crate_path::#label_type #ty_generics for #ident #where_clause {
```
Comment (Ratysz) [inline crates/bevy_ecs/macros/src/lib.rs:506]:
```suggestion
        impl #impl_generics #crate_path::#label_type for #ident #ty_generics #where_clause {
```
See [`syn` docs](https://docs.rs/syn/1.0.60/syn/struct.Generics.html#method.split_for_impl). Maybe there should be a test for this.

**Comment by aevyrie** [comment]
With the most recent change (and a lot of derives 🙂) this seems to fix the type errors.

However, from a useability standpoint, I'm finding it impossible to get this working due to all the trait bounds. My goal is to use my components as types in generic systems. However, when multiple generic systems are run that need labels, my system dependencies are between systems of the same type - which is what I'm trying to express. Is there some way I could wrap my components in something like a `Box<>` that implements all the `SystemLabel` functionality, without my components exploding with trait bound complexity?

**Comment by TheRawMeatball** [comment]
Hmm, can you post a snippet?

**Comment by aevyrie** [comment]
> Hmm, can you post a snippet?

Here's an example: https://github.com/aevyrie/bevy_mod_bounding/blob/9f5a1169fd742f58bac342c883df2001881594a0/src/lib.rs#L17-L25

I pushed the broken code to master.

**Comment by aevyrie** [comment]
My issue has been resolved here: https://github.com/bevyengine/bevy/issues/1497

**Comment by Ratysz** [comment]
> > Hmm, can you post a snippet?
> 
> Here's an example: https://github.com/aevyrie/bevy_mod_bounding/blob/9f5a1169fd742f58bac342c883df2001881594a0/src/lib.rs#L17-L25
> 
> I pushed the broken code to master.

Changing
```rust
PhantomData<T>: 'static + Clone + Hash + Debug + Eq + Send + Sync + BoundingVolume,
```
to
```rust
T: 'static + Clone + Hash + Debug + Eq + Send + Sync + BoundingVolume,
```
allows this to compile.

**Pull request closed without merging**